bpfassets/libbpf/src/kepler.bpf.c

```diff
@@ -198,13 +241,27 @@ struct sched_switch_info {
 SEC("tp/sched/sched_switch")
 int kepler_sched_switch_trace(struct sched_switch_info *ctx)
 {
-	u32 prev_pid, tgid, cpu_id;
-	u64 pid_tgid, cgroup_id, cur_ts;
-	pid_t cur_pid;
+	u32 prev_pid, next_pid, cpu_id;
+	u64 *prev_tgid;
+	long prev_state;
+	u64 cur_ts = bpf_ktime_get_ns();
 
-	struct process_metrics_t *cur_pid_metrics, *prev_pid_metrics;
+	struct process_metrics_t *prev_tgid_metrics, *curr_tgid_metrics;
 	struct process_metrics_t buf = {};
 
+	prev_state = ctx->prev_state;
+	prev_pid = (u32)ctx->prev_pid;
+	next_pid = (u32)ctx->next_pid;
+	cpu_id = bpf_get_smp_processor_id();
+
+	// Collect metrics
+	// Regardless of skipping the collection, we need to update the hardware
+	// counter events to keep the metrics map current.
+	collect_metrics_and_reset_counters(&buf, prev_pid, cur_ts, cpu_id);
```

Comment:
I'm concerned about the performance impact here. It potentially negates the benefit of sampling at all (which is still relatively expensive anyway).

Have you profiled the older version + this version so we can see how much CPU time is spent where? Map reads/writes are the most expensive operations and if we can't avoid those then I would expect sampling to be pointless.

Reply:
While skipping the counter update may seem beneficial for performance, it would lead to inaccurate data. By neglecting to update the counter, we would misattribute skipped threads metrics to subsequent thread. This would result in a significant overestimation of resource usage for individual threads.

Additionally, our current default deployment does not take advantage of sampling, and as you mentioned, more in-depth analysis is needed before employing it.